gpt_index/indices/response/response_synthesis.py

```diff
@@ -0,0 +1,179 @@
+
+   
+import logging
+from typing import (
+    Any,
+    Dict,
+    Generator,
+    List,
+    Optional,
+)
+
+from langchain.input import print_text
+
+from gpt_index.data_structs.node_v2 import Node, NodeWithScore
+from gpt_index.indices.query.schema import QueryBundle
+from gpt_index.indices.response.response_builder import (
+    BaseResponseBuilder,
+    BaseResponseBuilder,
+    ResponseMode,
+    get_response_builder,
+)
+from gpt_index.indices.service_context import ServiceContext
+from gpt_index.optimization.optimizer import BaseTokenUsageOptimizer
+from gpt_index.prompts.default_prompt_selectors import DEFAULT_REFINE_PROMPT_SEL
+from gpt_index.prompts.default_prompts import DEFAULT_SIMPLE_INPUT_PROMPT, DEFAULT_TEXT_QA_PROMPT
+from gpt_index.prompts.prompts import QuestionAnswerPrompt, RefinePrompt, SimpleInputPrompt
+from gpt_index.response.schema import (
+    RESPONSE_TYPE,
+    Response,
+    StreamingResponse,
+)
+from gpt_index.types import RESPONSE_TEXT_TYPE
+from gpt_index.utils import truncate_text
+
+logger = logging.getLogger(__name__)
+
+
+class ResponseSynthesizer:
+    def __init__(
+        self, 
+        response_builder: BaseResponseBuilder,
+        response_mode: ResponseMode,
+        response_kwargs: Optional[Dict] = None,
+        optimizer: Optional[BaseTokenUsageOptimizer] = None,
+        verbose: bool = False,
+    ) -> None:
```

Comment:
is response synthesizer the class that will be exposed to users? or also response builder? i'm assuming the former? 

(seems like the diff is just abstracting away the subclasses / adding optimizer?)

Reply:
Yes, the difference is pretty minor right now. `ResponseBuilder` is pretty generic, more just QA over text chunks. The `ResponseSynthsizer` adds additional features (like optimizer, but perhaps others in the future).